thirdparty/patchmatchnet/models/patchmatch.py

```diff
@@ -0,0 +1,951 @@
+
+        # sort
+        propagate_depth, _ = torch.sort(propagate_depth, dim=1)
+
+        return propagate_depth
+
+
+class Evaluation(nn.Module):
+    """Evaluation Class"""
+
+    def __init__(self, G: int = 8, stage: int = 3, evaluate_neighbors: int = 9, iterations: int = 2) -> None:
+        """Initialize method
+
+        Args:
+            G: number of groups
+            stage: stage id
+            evaluate neighbors: number of evaluate neighbors
+            iterations: number of evaluation iteration
+        """
+        super(Evaluation, self).__init__()
+
+        self.iterations = iterations
+
+        self.G = G
+        self.stage = stage
+        if self.stage == 3:
+            self.pixel_wise_net = PixelwiseNet(self.G)
+
+        self.similarity_net = SimilarityNet(self.G, evaluate_neighbors)
+
+    def forward(
+        self,
+        ref_feature: torch.Tensor,
+        src_features: List[torch.Tensor],
+        ref_proj: torch.Tensor,
+        src_projs: List[torch.Tensor],
+        depth_sample: torch.Tensor,
+        depth_min: int,
+        depth_max: int,
+        iter: int,
+        grid: torch.Tensor = None,
+        weight: torch.Tensor = None,
+        view_weights: torch.Tensor = None,
+    ) -> Union[Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]:
+        """Forward method for adaptive evaluation
+
+        Args:
+            ref_feature: feature from reference view
+            src_features: features from source views
+            ref_proj: projection matrix of reference view
+            src_projs: source matrices of source views
+            depth_sample: sample depth map,
+            depth_min: minimum virtual depth,
+            depth_max: maximum virtual depth,
+            iter: iteration number,
+            grid: grid,
+            weight: weight,
+            view_weights: list to store weights from source views,
+
+        Returns:
+            depth_sample: expectation of depth sample,
+            score: probability map,
+            view_weights: optional, similarity of images of different views
+        """
+        num_src_features = len(src_features)
+        num_src_projs = len(src_projs)
+        batch, feature_channel, height, width = ref_feature.size()
+        device = ref_feature.device
+
+        num_depth = depth_sample.size()[1]
+        assert (
+            num_src_features == num_src_projs
+        ), "Patchmatch Evaluation: Different number of images and projection matrices"
+        if view_weights is not None:
+            assert (
+                num_src_features == view_weights.size()[1]
+            ), "Patchmatch Evaluation: Different number of images and view weights"
+
+        pixel_wise_weight_sum = 0
+
+        ref_feature = ref_feature.view(batch, self.G, feature_channel // self.G, height, width)
+
+        similarity_sum = torch.Tensor([0])
+
+        if self.stage == 3 and view_weights is None:
+            view_weights_list = []

```

Comment:
was there any particular reason to rename `view_weights` to `view_weights_list` here? for mypy?

Reply:
`view_weights` (when not `None` in input) should be `torch.Tensor`. Here `view_weights` is initialized as `List`.